reader-osm/src/main/java/com/graphhopper/reader/osm/OSMReader.java

```diff
@@ -573,10 +573,30 @@ int addTowerNode(long osmId, double lat, double lon, double ele) {
         final PointList pointList = new PointList(osmNodeIds.size(), nodeAccess.is3D());
         final List<EdgeIteratorState> newEdges = new ArrayList<>(5);
         int firstNode = -1;
-        final int lastIndex = osmNodeIds.size() - 1;
         int lastInBoundsPillarNode = -1;
         try {
-            for (int i = 0; i < osmNodeIds.size(); i++) {
+            // #2221: ways might include nodes at the beginning or end that do not exist -> skip them
+            int firstExisting = -1;
+            int lastExisting = -1;
```

Comment:
The actual fix is a bit lengthy, but rather trivial: We scan the osmNodeIds list from the beginning and end to find the first node that does exist (and skip those that do not exist).